### PR TITLE
Added regex support

### PR DESCRIPTION
**Pull request by idan-h**
Tested pr, quite simple, that adds regex support for the ignore, require, and replace functionalities.

It works  by adding `$regex:` to the start of each string, and will else function as usual.

Might be worth adding some example patterns, maybe to the readme or something. Not a must.

An example pattern to insert the word space between numbers:
`$regex:(?<=\d)\s+(?=\d)` replace with `space`

An example pattern to insert spaces between numbers:
`$regex:(?<=[0-9])(?=[0-9])` replace with ` `

it is also possible to use group references by just typing $1, $2 etc.

### CODE REVIEW

**Review by pilot51 — CHANGES_REQUESTED** [review]
Thanks for the contribution!

In addition to the requested code changes, the descriptions for all 3 options should mention how to enable regex. For now we can assume the user knows or can figure out for themselves how to write regex, they just need to be made aware of how to enable it.

Strings:
`require_ignore_strings_dialog_msg`
`tts_text_replace_dialog`

File: app/src/main/java/com/pilot51/voicenotify/NotificationInfo.kt
Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/NotificationInfo.kt:137]: I think that regex should be always case-sensitive. I added this in strings.
I am not sure how to implement the toast thing, but I fixed the crash

File: app/src/main/java/com/pilot51/voicenotify/Service.kt
Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:233]: done

File: app/src/main/java/com/pilot51/voicenotify/Service.kt
Comment (pilot51) [inline app/src/main/java/com/pilot51/voicenotify/Service.kt:250]: I'm not really a fan of creating local functions like this. I think moving it to the companion object as an extension would be better:
```
private fun String.containsOrMatchesRegex(it: String): Boolean {
	return if (it.startsWith(Constants.REGEX_PREFIX)) {
		try {
			Regex(it.removePrefix(Constants.REGEX_PREFIX)).containsMatchIn(this)
		} catch (e: PatternSyntaxException) {
			e.printStackTrace()
			CoroutineScope(Dispatchers.Main).launch {
				Toast.makeText(appContext, R.string.malformed_regex, Toast.LENGTH_LONG).show()
			}
			false
		}
	} else contains(it, true)
}
```

Usage: `ttsMsg.containsOrMatchesRegex(it)`

**Review by pilot51 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/pilot51/voicenotify/Service.kt
Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:250]: Well, local functions are there for a reason - they are releveant only to the code using them. We could just duplicate the code and ditch the function, but I think it is less elegant.

Also, the toast should be only when saving the regex no? There is not reason to put it on the service

Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:250]:
```
I agree, case sensitive allows for more flexibility and I'm sure anyone who wants to use regex would appreciate it.
Unfortunately, it would appear you forgot to update the code in both places when you changed the strings to say case sensitive.
```
Sorry, but I didn't understand what I forgot, could you please refer me to the file?

Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:250]: I believe that everything has it's time and place, and sometimes nested functions might be the best solution, like in this case.
I will change it to lambda.
You are welcome to refactor if you would like, I have no problem with that. I just think that a code refactor should be backed by a good reason, and I don't really see the benefit here.

I will try to move it on save, but I am a bit short on time so if I get stuck I will use the lazy solution.

I saw, thanks

Comment (pilot51) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:250]: It's just a matter of familiarity. I don't think I've seen anyone use a local function before even though I work with other experienced Kotlin devs and there's plenty of opportunity to do so in those much larger projects, so it looks weird to me. Maybe that's because we all come from a Java background and it's almost never mentioned in Kotlin documentation and examples. I haven't given it much thought until now, so I'm looking around for other perspectives from the larger Kotlin community and I may change my mind. Feel free to leave it a local function if you prefer.

On the other hand, Kotlin allows functions and properties at the file level outside of any class, again unlike Java. It felt wrong at first, but thanks to how common it is for Compose, I've gotten used to it.

Comment (idan-h) [reply app/src/main/java/com/pilot51/voicenotify/Service.kt:250]: I actually come from a c# background, I don't really know Kotlin 😄

I rarely use local functions aswell, but on some occasions they are helpful.

**Comment by pilot51** [comment]
I know you've been busy, but I plan to merge this weekend and make the changes if you don't.

The important one is removing `RegexOption.IGNORE_CASE` at Service:254.
Validation on save would also be great but not critical.

I plan to add some useful examples to the string after merge to satisfy some of the requests I've been getting such as filtering out URLs. I would add the blank line when I do that, so don't worry about it. In fact, I may make the regex paragraph a separate string and append it to the rest of the description to eliminate the duplication.

**Comment by idan-h** [comment]
> I know you've been busy, but I plan to merge this weekend and make the changes if you don't.
> 
> The important one is removing `RegexOption.IGNORE_CASE` at Service:254. Validation on save would also be great but not critical.
> 
> I plan to add some useful examples to the string after merge to satisfy some of the requests I've been getting such as filtering out URLs. I would add the blank line when I do that, so don't worry about it. In fact, I may make the regex paragraph a separate string and append it to the rest of the description to eliminate the duplication.

Hey, if you didn't I will probably get to it tommorow

**Comment by idan-h** [comment]
I made the small change with the regex, unfortunately I ran out of time for the toast, but I added an unharmful commit so you can continue

**Pull request merged**
